// mapv-three/dist/addons/traffic/config/assets.js
/*
 * @Author: wangchen35
 * @Date: 2023-01-04 18:54:04
 * @LastEditors: wangchen35
 * @LastEditTime: 2023-01-12 17:14:18
 * @Description: file content
 */
export const TEMPLTE_MODEL_CONFIG = {
    x: 0,
    y: 0,
    z: 0,
    color: '',
    name: '',
    rotateX: 0,
    rotateY: 0,
    rotateZ: 0,
    scaleX: 1,
    scaleY: 1,
    scaleZ: 1,
    scale: 1,
    opacity: 1,
    isTips: false,
    isVisible: true,
    isClick: true,
    isBloom: 0,
    classType: '',
    lodMax: 1,
    lodMin: 1,
    text: '',
    // 是否进行颜色校正，默认需要
    encoding: 1,
    reflectivity: 0,
    roughness: 0,
    metalness: 0,
    modelName: '全部',
    modelChildName: [],
};

export const TEMPLATE_POINT = {
    uuid: '',
    project_id: '',
    x: 0,
    y: 0,
    z: 0,
    name: '',
    rotate_x: 0,
    rotate_y: 0,
    rotate_z: 0,
    scale_x: 1,
    scale_y: 1,
    scale_z: 1,
    scale: 1,
    is_tips: 0,
    is_visible: 1,
    is_click: 1,
    class_type: '',
    lod_max: 1,
    lod_min: 1,
    opacity: '',
    text: '',
    bounding_box: '',
    has_bounding_box: 0,
    layer_type: '',
    layer_name: '',
    layer_id: 1,
    model_name: '',
    status: 1,
    is_bloom: 0,
    material_id: '',
    material_list: '',
    images: '[]',
    model_info: '[]',
    device_id: '',
};
